rfcs/0004-categorized-attributes.md

```diff
@@ -0,0 +1,217 @@
+
+There MUST be a _default modifier_ indicating the base value of a non over-time effect.
+
+## Data format
+
+The data format MUST define a keyword for every _category_. It MUST either define a keyword for every _effect_ and _modifier_ individually, or a keyword for every _effect_ and _modifier_ combination. It MUST NOT define a keyword for the _default modifier_.
+
+These keywords MUST be supported for parsing, and MUST NOT be context-sensitive. These keywords MUST be supported by a parser in a nested approach.
+
+### Parsers
+
+A parser MAY follow one of the example formats:
+
+```
+ship <name>
+	attributes
+		<category-keyword> [<value#>]
+			<effect-keyword> [<value#>]
+				<modifier-keyword> [<value#>]
+```
+
+```
+ship <name>
+	attributes
+		<category-keyword> [<value#>]
+			<effect-modifier-keyword> [<value#>]
+```
+
+A parser MUST support parsing from a ship, outfit or weapon definition. It MUST support parsing attributes from `attribute` nodes containing mixed _new-form_ and _old-form_ _categorized attributes_. A parser MUST support all _old-form_ and _new-form_ attribute definitions. A parser MAY provide feedback about unsupported attributes.
+
+A parser SHALL provide feedback about unsupported value tokens. It MUST support a value token for the _effect_ node, implying the _default modifier_. It MAY support values for _category_ keywords directly, implying a default _effect_. (This _effect_ SHOULD be different for every _category_.) A parser SHALL support applicable _category_, _effect_ and _modifier_ nodes without a value, assuming a default value of 1. This behaviour MUST NOT be conditional on the presence of child nodes.
+
+A parser MUST abstract the presence of _old-form_ _categorized attributes_ from the engine.
+
+### Generators
+
+A generator for the save file format MUST support _new-form_ attribute generation. It SHALL NOT generate _old-form_ syntax for _categorized attributes_. It MUST NOT generate duplicate entries of the _category_, _effect_, _modifier_ or combined _modifier_ + _effect_ keywords.
+
+A generator SHALL use a data node to represent a _category_ and _effect_ or _effect_ + _category_ combination. A generator SHOULD NOT include attributes with their default value.
+
+A generator SHOULD use a deterministic output ordering for attributes, and MAY order them alphabetically. It SHALL provide the most concise output possible.
+
+A generator MAY NOT support _old-form_ _categorized attributes_.
+
+### Engine support
+
+The supported _categories_ SHOULD include all of the following:
+
+```
+shield generation
+hull repair
+thrusting
+reverse thrusting
+steering
+active cooling
+ramscooping
+cloaking
+afterburning
+firing
+```
+
+It MAY include any or all of the following:
+```
+protection
+resistance
+damage
+capacity
+```
+
+The supported _effects_ SHOULD include all of the following:
+```
+shields
+hull
+thrust
+reverse thrust
+turn
+active cooling
+ramscoop
```

Comment:
I don't understand what those effects mean, can a weapon remove thrust, turn power or ramscoop efficiency? (The actual benefits provided by an outfit were already handled 1 level higher.)

Reply:
So we can have weapons that causes a ship to turn when fired? That is actually pretty cool. Does it also work for damage (causing a target ship to turn)?

I'm still not sure about `active cooling`, isn't that a category that provides `cooling`?

Reply:
Turning will need some special handling, as there is no direction specified, and the "turn" attributes for engines only specify potential turn, but it could be done. (You can, right now, have a weapon that provides "turn" like an engine, but not one that makes the ship turn when fired.)

The main difference between `active cooling` and `cooling` is how they are applied; active cooling scales with the heat level of the ship, which iirc also affects the energy consumption of active cooling outfits? I'm not sure if we can merge them without breaking functionality, but it's something to keep in mind. It might be possible.